File: frontend/lib/permissions.js
// lib/permissions.js

import { USER_ROLES } from '@/lib/constants';

const ROLE_HIERARCHY = {
  [USER_ROLES.SUPER_ADMIN]: 50,
  [USER_ROLES.OWNER]: 40,
  [USER_ROLES.ADMIN]: 30,
  [USER_ROLES.DEVELOPER]: 20,
  [USER_ROLES.VIEWER]: 10,
};

export const PERMISSIONS = {
  PROJECTS_CREATE: 'projects:create',
  PROJECTS_DELETE: 'projects:delete',
  PROJECTS_UPDATE: 'projects:update',
  PROJECTS_VIEW: 'projects:view',

  AGENTS_CREATE: 'agents:create',
  AGENTS_DELETE: 'agents:delete',
  AGENTS_RUN: 'agents:run',
  AGENTS_VIEW: 'agents:view',

  KNOWLEDGE_BASES_CREATE: 'knowledge_bases:create',
  KNOWLEDGE_BASES_DELETE: 'knowledge_bases:delete',
  KNOWLEDGE_BASES_UPDATE: 'knowledge_bases:update',
  KNOWLEDGE_BASES_VIEW: 'knowledge_bases:view',

  VECTOR_STORES_MANAGE: 'vector_stores:manage',
  VECTOR_STORES_VIEW: 'vector_stores:view',

  MONITORING_VIEW: 'monitoring:view',

  SETTINGS_MANAGE: 'settings:manage',
  SETTINGS_API_KEYS: 'settings:api_keys',
  SETTINGS_USERS: 'settings:users',
  SETTINGS_SECURITY: 'settings:security',

  BILLING_MANAGE: 'billing:manage',
};

const ROLE_PERMISSIONS = {
  [USER_ROLES.SUPER_ADMIN]: Object.values(PERMISSIONS),
  [USER_ROLES.OWNER]: Object.values(PERMISSIONS),
  [USER_ROLES.ADMIN]: [
    PERMISSIONS.PROJECTS_CREATE,
    PERMISSIONS.PROJECTS_UPDATE,
    PERMISSIONS.PROJECTS_VIEW,
    PERMISSIONS.AGENTS_CREATE,
    PERMISSIONS.AGENTS_DELETE,
    PERMISSIONS.AGENTS_RUN,
    PERMISSIONS.AGENTS_VIEW,
    PERMISSIONS.KNOWLEDGE_BASES_CREATE,
    PERMISSIONS.KNOWLEDGE_BASES_DELETE,
    PERMISSIONS.KNOWLEDGE_BASES_UPDATE,
    PERMISSIONS.KNOWLEDGE_BASES_VIEW,
    PERMISSIONS.VECTOR_STORES_MANAGE,
    PERMISSIONS.VECTOR_STORES_VIEW,
    PERMISSIONS.MONITORING_VIEW,
    PERMISSIONS.SETTINGS_API_KEYS,
    PERMISSIONS.SETTINGS_USERS,
  ],
  [USER_ROLES.DEVELOPER]: [
    PERMISSIONS.PROJECTS_VIEW,
    PERMISSIONS.AGENTS_CREATE,
    PERMISSIONS.AGENTS_RUN,
    PERMISSIONS.AGENTS_VIEW,
    PERMISSIONS.KNOWLEDGE_BASES_CREATE,
    PERMISSIONS.KNOWLEDGE_BASES_UPDATE,
    PERMISSIONS.KNOWLEDGE_BASES_VIEW,
    PERMISSIONS.VECTOR_STORES_VIEW,
    PERMISSIONS.MONITORING_VIEW,
    PERMISSIONS.SETTINGS_API_KEYS,
  ],
  [USER_ROLES.VIEWER]: [
    PERMISSIONS.PROJECTS_VIEW,
    PERMISSIONS.AGENTS_VIEW,
    PERMISSIONS.KNOWLEDGE_BASES_VIEW,
    PERMISSIONS.VECTOR_STORES_VIEW,
    PERMISSIONS.MONITORING_VIEW,
  ],
};

export function hasPermission(role, permission) {
  if (!role || !permission) return false;
  const permissions = ROLE_PERMISSIONS[role] || [];
  return permissions.includes(permission);
}

export function hasAnyPermission(role, permissionList = []) {
  return permissionList.some((permission) => hasPermission(role, permission));
}

export function hasAllPermissions(role, permissionList = []) {
  return permissionList.every((permission) => hasPermission(role, permission));
}

export function isRoleAtLeast(role, minimumRole) {
  const roleLevel = ROLE_HIERARCHY[role] ?? 0;
  const minimumLevel = ROLE_HIERARCHY[minimumRole] ?? 0;
  return roleLevel >= minimumLevel;
}

export function canManageWorkspace(role) {
  return isRoleAtLeast(role, USER_ROLES.ADMIN);
}

export function canManageBilling(role) {
  return isRoleAtLeast(role, USER_ROLES.OWNER);
}

File: backend/multitenancy/organizations/permissions.py
# multitenancy/organizations/permissions.py
from __future__ import annotations

import enum

from settings import Role


class Permission(str, enum.Enum):
    ORG_VIEW = "organization:view"
    ORG_UPDATE = "organization:update"
    ORG_DELETE = "organization:delete"
    MEMBERS_VIEW = "members:view"
    MEMBERS_INVITE = "members:invite"
    MEMBERS_REMOVE = "members:remove"
    MEMBERS_UPDATE_ROLE = "members:update_role"
    BILLING_VIEW = "billing:view"
    BILLING_MANAGE = "billing:manage"
    PROJECTS_CREATE = "projects:create"
    PROJECTS_VIEW = "projects:view"
    PROJECTS_UPDATE = "projects:update"
    PROJECTS_DELETE = "projects:delete"
    AGENTS_CREATE = "agents:create"
    AGENTS_EXECUTE = "agents:execute"
    AGENTS_DELETE = "agents:delete"
    KNOWLEDGE_BASES_MANAGE = "knowledge_bases:manage"
    DATASETS_MANAGE = "datasets:manage"
    WORKFLOWS_MANAGE = "workflows:manage"
    API_KEYS_MANAGE = "api_keys:manage"
    AUDIT_LOG_VIEW = "audit_log:view"
    QUOTAS_MANAGE = "quotas:manage"


_VIEWER_PERMS: frozenset[Permission] = frozenset({
    Permission.ORG_VIEW,
    Permission.MEMBERS_VIEW,
    Permission.PROJECTS_VIEW,
})

_DEVELOPER_PERMS: frozenset[Permission] = _VIEWER_PERMS | frozenset({
    Permission.PROJECTS_CREATE,
    Permission.PROJECTS_UPDATE,
    Permission.AGENTS_CREATE,
    Permission.AGENTS_EXECUTE,
    Permission.KNOWLEDGE_BASES_MANAGE,
    Permission.DATASETS_MANAGE,
    Permission.WORKFLOWS_MANAGE,
    Permission.API_KEYS_MANAGE,
})

_ADMIN_PERMS: frozenset[Permission] = _DEVELOPER_PERMS | frozenset({
    Permission.ORG_UPDATE,
    Permission.MEMBERS_INVITE,
    Permission.MEMBERS_REMOVE,
    Permission.MEMBERS_UPDATE_ROLE,
    Permission.PROJECTS_DELETE,
    Permission.AGENTS_DELETE,
    Permission.AUDIT_LOG_VIEW,
    Permission.QUOTAS_MANAGE,
})

_OWNER_PERMS: frozenset[Permission] = _ADMIN_PERMS | frozenset({
    Permission.ORG_DELETE,
    Permission.BILLING_VIEW,
    Permission.BILLING_MANAGE,
})

_SUPER_ADMIN_PERMS: frozenset[Permission] = frozenset(Permission)

ROLE_PERMISSIONS: dict[Role, frozenset[Permission]] = {
    Role.VIEWER: _VIEWER_PERMS,
    Role.DEVELOPER: _DEVELOPER_PERMS,
    Role.ADMIN: _ADMIN_PERMS,
    Role.OWNER: _OWNER_PERMS,
    Role.SUPER_ADMIN: _SUPER_ADMIN_PERMS,
}


def has_permission(role: Role, permission: Permission) -> bool:
    return permission in ROLE_PERMISSIONS.get(role, frozenset())


def permissions_for_role(role: Role) -> frozenset[Permission]:
    return ROLE_PERMISSIONS.get(role, frozenset())